mlflow/keras.py

```diff
@@ -461,21 +467,43 @@ def fit(self, *args, **kwargs):
         unlogged_params = ['self', 'x', 'y', 'callbacks', 'validation_data', 'verbose']
 
         log_fn_args_as_params(original, args, kwargs, unlogged_params)
+        early_stop_callback = None
 
         # Checking if the 'callback' argument of fit() is set
         if len(args) >= 6:
             tmp_list = list(args)
+            early_stop_callback = early_stop_check(tmp_list[5])
             tmp_list[5] += [__MLflowKerasCallback()]
             args = tuple(tmp_list)
         elif 'callbacks' in kwargs:
+            early_stop_callback = early_stop_check(kwargs['callbacks'])
             kwargs['callbacks'] += [__MLflowKerasCallback()]
         else:
             kwargs['callbacks'] = [__MLflowKerasCallback()]
 
-        result = original(self, *args, **kwargs)
+        history = original(self, *args, **kwargs)
+
+        # Checking if an EarlyStopping callback was used
+        if early_stop_callback:
+            stopped_epoch = early_stop_callback.stopped_epoch
+            patience = early_stop_callback.patience
+            try_mlflow_log(mlflow.log_metric, 'stopped_epoch', stopped_epoch)
+            try_mlflow_log(mlflow.log_param, 'earlystopping_patience', patience)
+            # Weights are restored only if early stopping occurs
+            if stopped_epoch != 0 and early_stop_callback.restore_best_weights:
+                best_epoch = stopped_epoch - max(1, patience)
+                try_mlflow_log(mlflow.log_metric, 'best_epoch', best_epoch)
+                best_metrics = {key: history.history[key][best_epoch]
+                                for key in history.history.keys()}
+                # Checking that a metric history exists
+                metric_key = next(iter(history.history), None)
+                if metric_key is not None:
+                    last_epoch = len(history.history[metric_key])
+                    try_mlflow_log(mlflow.log_metrics, best_metrics, step=last_epoch)
```

Comment:
I think these lines should be put in a function so that we can reuse it on fit_generator.

Reply:
Good idea! Will do that.